sysctl: Rename ignoreerrors to ignore_unknown_keys

winem opened the pull request:
This PR renames ignoreerrors to ignore_errors and keeps backward compatibility with ignorerrors. 

It's basically to be consistent with all other modules.

The sysctl module is the only one where the attribute is called ignoreerrors and not ignore_errors.
The new parameter is called ignore_errors but supports ignoreerrors as an alias.
The alias ensures that backward compatibility is given and this is not a breaking change.

##### SUMMARY
With this PR ignore_errors is allowed, too. I found it a bit unintuitive because it is not consistent with all other modules where the parameter is called ignore_errors with an underscore. The alias is to ensure backward compatibility because this cosmetically change should not break any playbooks out there. 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
sysctl

##### ADDITIONAL INFORMATION
Documentation was updated and 2 tests to test both, ignoreerrors and ignore_errors were added.

Commit:
Rename ignoreerrors to ignore_errors

This PR is basically to be consistent with all other modules.

The sysctl module is the only one where the attribute is called ignoreerrors and not ignore_errors.

The new parameter is called ignore_errors but supports ignoreerrors as an alias.

The alias ensures that backward compatibility is given and this is not a breaking change.

bcoca commented:
modules don't implement ignore_errors, it is a keyword of the task, this will lead to confusion, i almost prefer keeping the old name to disambiguate.

bcoca reviewed (changes requested):
modules don't implement ignore_errors, it is a keyword of the task, this will lead to confusion, i almost prefer keeping the old name to disambiguate.

Commit:
add debug task for ignore_error tests

winem commented:
@bcoca I totally agree here. It's a bit confusing especially when you combine both or use both in the same playbooks. To be honest, it even took me a moment to understand the different effects of ignoreerrors and ignore_errors. 

So I took another look at the other modules and it's really just the keyword that it's referenced there. 

I have mixed feelings about this PR and am not sure if consistency is worth the risk for confusion here. I let this up to you, guys. I also don't mind to change anything here. 

(Will take a look at the failing tests for Fedora, CentOS, etc. tomorrow. Already 3 a.m. here.)

winem commented:
I thought a bit more about this and, if at all, it probably just makes sense to choose another name for the parameter (i.e. ignore_unknown_keys). ignore_errors on 2 different levels is something that should be avoided. 

The reason why I changed my opinion a bit is this from the previous comment: "So I took another look at the other modules and it's really just the keyword that it's referenced there." 

If there is no further input or request from anyone else with an opinion on this toic, I'll close this PR.

winem commented:
@bcoca what do you think about the option to rename it to "ignore_unknown_keys"? I think that this is a small improvement and not as ambiguous as `ignoreerrors` (although the documentation is unambiguous here.
So this is more for people just reading a playbook without having all the modules and parameters in mind.

gundalow commented:
`ignore_unknown_keys` sounds good

winem commented:
Ok, will provide an updated PR.

Commit:
sysctl - rename ignoreerrors to ingore_unknown_keys

ignore_unknown_keys is not as ambiguous as ignorerrors
ignoreerrors could be confused with the task keyword ignore_errors
and ignore_unknown_keys is more specific about in regards to what it actually does

mattclay commented:
Some of our test containers don't support reloading. Using `reload: no` as is done on the other tasks will work. Is there a reason why that can't be done for the new tests?

bcoca commented:
@mattclay can we make it conditional? it would be nice to also test reloading .. since it is an integral part of the plugin

winem commented:
I'm all in for conditionals on the conditional reload. Just let me know the OS or rather OS family (or any other condition) for containers that do not support reload. Any details are appreciated. @mattclay

winem commented:
Ready for review and further input. @mattclay your input is highly appreciated.

ready_for_review

mattclay commented:
@winem Probably the best you can do is to call `sysctl` to set a parameter and see if it fails. If it does, then avoid reload during the test. I expect that should allow testing reload on our RHEL VMs, while skipping on most (if not all) of our Docker containers.

winem commented:
@mattclay here is my Christmas gift for you. 

I added a check to see if the sysctl command is available and we have a conditional on the tasks that I suspect to fail if it's not available. Please feel free to test it. My first tests look good. 

This affects just the tests where `sysctl_set: true` was used or the `sysctl` command was executed via the shell module. Let me know if I missed anything or you see further ways to improve the tests.

Commit:
use reload only if it's supported by the OS

winem commented:
@mattclay here is my Christmas gift for you.

I added a check to see if the sysctl command is available and use `reload: yes` only when it's available.

bcoca commented on lib/ansible/modules/system/sysctl.py:
```diff
@@ -241,13 +242,13 @@ def set_token_value(self, token, value):
         elif self.platform == 'freebsd':
             ignore_missing = ''
-            if self.args['ignoreerrors']:
+            if self.args['gnore_unknown_keys']:
```
missing an 'i'

winem replied:
@bcoca thanks, this was a mean one. Looks like a typical vi... 
Fixed will push it with the fix for systems not supporting sysctl.

Commit:
typo; missing i

Commit:
sysctl_set only if sysctl is available

Commit:
add missing debugs to increase visbility during tests

Commit:
change sysctl test to fit containers

Commit:
check if sysctl command available & execute ignore_unknown_keys unit tests only if os family != 'RedHat'

Commit:
debug...

Commit:
Revert "debug..." - it works as expected now

This reverts commit 611a8852396461bbbecd010bc41040d5382802f1.

winem commented:
For some reason, the RedHat-Container return a "read-only filesystem" error (I did not notice that it was related immediately, so I'm sorry for the delay. I thought we're talking about two issues here.)

Now I fixed it and we handle 2 cases: 
1. if the sysctl command is not available at all (I even changed the check from `command -v sysctl` to `sysctl -w ...` when troubleshooting the issue. 
2. we do not run the ignore_unknown_keys unit tests on RedHat. 

Let me know if this is ok for you. To me it feels "ok-ish" but I'm not sure how satisfied I am with that solution. 
 
ready_for_review

mattclay commented on test/integration/targets/sysctl/tasks/main.yml:
```diff
@@ -143,31 +168,80 @@
+    ignore_unknown_keys: yes
+    reload: "{{ sysctl_cmd_supported }}"
+  register: sysctl_ignore_unknown_keys0
+  when: ansible_os_family != 'RedHat'
```
This results in skipping the tests on the only Linux VM we use in CI, which makes me think this is simply covering up an issue. To make sure this is correct we need to understand why it's failing.

In the case of our Docker containers we know it's a permissions issue, but that shouldn't be the case on our RHEL virtual machines.

winem replied:
Totally agree and this is the part that does not really satisfy me. The test works fine on RedHat VMs and just fails on Docker Containers and (and that's the actual challenge I think) the response from the sysctl command in this case is not "key does not exist". Instead it was either read-only fs or something similiar. So the `sysctl -e` does not recognize it a invalid key and fails instead. 

I do not have an idea right now how to fix this or write a reliable and satisfying test here but I'm open for any input.

mattclay commented on test/integration/targets/sysctl/tasks/main.yml:
```diff
@@ -29,6 +29,28 @@
     state: directory
 
+- name: check if sysctl command is available
+  shell: sysctl -w net.ipv4.ip_forward=0
```
I recommend a setting less likely to interfere with regular system operation:

```suggestion
  shell: sysctl -w debug.exception-trace=1
```

It would also be a good idea to query the existing value and use that for the set, since we don't actually need to change the setting, just verify that we have permissions to do so.

Also, the test should have `needs/root` added to the aliases file, since it requires root permissions. Since we're potentially changing sysctl values, `destructive` should also be added.

winem replied:
Oh, the debug keys are great for such a test. Will update it.

mattclay commented on test/integration/targets/sysctl/tasks/main.yml:
```diff
@@ -125,15 +147,18 @@
     value: 1
     sysctl_set: yes
     reload: no
+  when: sysctl_cmd_supported
```
Why is this needed? The existing tests pass without it and should only be updating `sysctl.conf` without writing the new values with `sysctl`.

winem replied:
Here we test the sysctl_set function. This executes `sysclt -w` and fails on the docker containers, too. So we skip these tests if the first `sysctl -w` failed. This is again a bit "suboptimal" as it might fail for various reasons. If you have a better idea how to fix this, let me hear and learn. :)

mattclay commented on test/integration/targets/sysctl/tasks/main.yml:
```diff
@@ -29,6 +29,28 @@
+    verbosity: 1
+
+- set_fact:
+    sysctl_cmd_supported: yes
```
Wouldn't it be more accurate to call this `sysctl_write_supported`?

winem replied:
Yes, now, when I review it, I got lost between the "Some of our test containers don't support reloading" and the file-system issues where I did not see the correlation right from the beginning. Will wait for your feedback on the other comments and am happy to change this then.
